calyx-py/test/correctness/queues/fifo.py

```diff
@@ -27,75 +27,54 @@ def insert_fifo(prog, name, queue_len_factor=QUEUE_LEN_FACTOR):
+    # The user called pop/peek.
+    # If the queue is empty, we should raise an error.
+    # Otherwise, we should proceed with the core logic
+    pop_peek_logic = cb.if_with(
+        fifo.eq_use(len.out, 0),
+        raise_err,
+        [
+            # `pop` or `peek` has been called, and the queue is not empty.
+            fifo.mem_load_d1(mem, read.out, ans, "read_payload_from_mem"),
+            # Write the answer to the answer register.
+            # If the user called pop, increment `read` and decrement `len`.
+            cb.if_with(fifo.eq_use(cmd, 0), [fifo.incr(read), fifo.decr(len)]),
+        ],
+    )
+
+    # The user called push.
+    # If the queue is full, we should raise an error.
+    # Otherwise, we should proceed with the core logic.
+    push_logic = cb.if_with(
+        fifo.eq_use(len.out, max_queue_len),
+        raise_err,
+        [  # `push` has been called and the queue is not full.
+            # Write `value` to the queue, and increment `write` and `len`.
+            fifo.mem_store_d1(mem, write.out, value, "write_payload_to_mem"),
+            fifo.incr(write),
+            fifo.incr(len),
+        ],
+    )
 
     fifo.control += cb.par(
         # Was it a (pop/peek), or a push? We can do those two cases in parallel.
         # The logic is shared for pops and peeks, with just a few differences.
-        cb.if_with(
-            # Did the user call pop/peek?
-            cmd_lt_2,
-            cb.if_with(
-                # Yes, the user called pop/peek. But is the queue empty?
-                len_eq_0,
-                raise_err,  # The queue is empty: underflow.
-                [  # The queue is not empty. Proceed.
-                    write_to_ans,  # Write the answer to the answer register.
-                    cb.if_with(
-                        cmd_eq_0,  # Did the user call pop?
-                        [  # Yes, so we have some work to do besides peeking.
-                            read_incr,  # Increment the read pointer.
-                            len_decr,  # Decrement the active length.
-                        ],
-                    ),
-                ],
-            ),
-        ),
-        cb.if_with(
-            # Did the user call push?
-            cmd_eq_2,
-            cb.if_with(
-                # Yes, the user called push. But is the queue full?
-                len_eq_max_queue_len,
-                raise_err,  # The queue is empty: underflow.
-                [  # The queue is not full. Proceed.
-                    write_to_mem,  # Write `value` to the queue.
-                    write_incr,  # Increment the write pointer.
-                    len_incr,  # Increment the active length.
-                ],
-            ),
-        ),
+        # Did the user call pop/peek?
+        cb.if_with(fifo.lt_use(cmd, 2), pop_peek_logic),
+        # Did the user call push?
+        cb.if_with(fifo.eq_use(cmd, 2), push_logic),
```

Comment:
Unless I'm missing something, it seems like we're not raising the error flag when `cmd = 3` here?

Reply:
Yes, that's a lesson I ought to learn from you. Will do!

Reply:
Done in [bcee73e](https://github.com/calyxir/calyx/pull/2163/commits/bcee73ed9982094e3310fa9cd6d187e554dd4314)